drivers/power_domain/power_domain_device_onoff.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+#include <zephyr/pm/pm.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(power_domain_deviceonoff, CONFIG_POWER_DOMAIN_LOG_LEVEL);
+
+/* Indicates the end of the onoff_power_states array for count purposes */
+#define POWER_DOMAIN_DEVICE_ONOFF_STATE_COUNT	0xFF
+
+struct pd_deviceonoff_config {
+	uint8_t *onoff_power_states;
+};
+
+struct pd_visitor_context {
+	const struct device *domain;
+	enum pm_device_action action;
+};
+
+static int pd_domain_visitor(const struct device *dev, void *context)
+{
+	struct pd_visitor_context *visitor_context = context;
+	int ret = 0;
+
+	/* Only run action if the device is on the specified domain */
+	if (!dev->pm || (dev->pm->domain != visitor_context->domain)) {
+		return 0;
+	}
+
+	/* In case device is active, first suspend it before turning it off */
+	if ((visitor_context->action == PM_DEVICE_ACTION_TURN_OFF) &&
+		(dev->pm->state == PM_DEVICE_STATE_ACTIVE)) {
+		ret = pm_device_action_run(dev, PM_DEVICE_ACTION_SUSPEND);
+		/* Check the type of failure */
+		if (ret != 0) {
+			if ((ret != -ENOSYS) || (ret != -ENOTSUP) || (ret != -EALREADY)) {
+				return ret;
+			}
+		}
+	}
+	(void)pm_device_action_run(dev, visitor_context->action);
+	return 0;
+}
+
+static int pd_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct pd_deviceonoff_config *config = dev->config;
+	uint8_t i = 0;
+	int ret = 0;
+	enum pm_state state = pm_state_next_get(0)->state;
+	struct pd_visitor_context context = {.domain = dev};
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		while (config->onoff_power_states[i] != POWER_DOMAIN_DEVICE_ONOFF_STATE_COUNT) {
+			if (state == config->onoff_power_states[i]) {
+				/* Notify devices on the domain they are now powered */
+				context.action = PM_DEVICE_ACTION_TURN_ON;
+				ret = device_supported_foreach(dev, pd_domain_visitor, &context);
+				/* No need to go through the rest of the array of states */
+				break;
+			}
+			i++;
+		}
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		while (config->onoff_power_states[i] != POWER_DOMAIN_DEVICE_ONOFF_STATE_COUNT) {
+			if (state == config->onoff_power_states[i]) {
+				/* Notify devices on the domain that power is going down */
+				context.action = PM_DEVICE_ACTION_TURN_OFF;
+				ret = device_supported_foreach(dev, pd_domain_visitor, &context);
+				/* No need to go through the rest of the array of states */
+				break;
+			}
+			i++;
+		}
+		break;
+	case PM_DEVICE_ACTION_TURN_ON:
+		break;
+	case PM_DEVICE_ACTION_TURN_OFF:
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return ret;
```

Comment:
I don't think this will work as you expect since the pm subsystem will try to suspend devices before the domain (if they are hierarchically  correct), then when the pm action is called for this domain, all devices under it will already be suspended.

Another case this can cause issue is if one enable device runtime pm for this device. In this case, if one does `pm_device_runtime_put(domain_device_on_off);` while the system is active, `pm_state_next_get(0)->state;` will return `PM_STATE_ACTIVE` and it will much likely not suspend devices.

Please take a look in 
https://github.com/zephyrproject-rtos/zephyr/pull/70623 I have tried to address this problem where only specifics power states should trigger device pm.

Reply:
@ceolin, I have update the implementation based on PR https://github.com/zephyrproject-rtos/zephyr/pull/74371